### ISSUE TITLE
devkit's `readTargetOptions` doesn't consider targets infered from `registerProjectTargets` in plugins

### ISSUE THREAD

**Issue by JulioC** [nrwl/nx#10972]
## Current Behavior

When calling `readTargetOptions`, it uses `context.workspace.projects` to get the target object, but this property is not being populated with targets registered using `registerProjectTargets` from addons.

This is an issue when implicitly defining a target for `@nrwl/web:dev-server`, which uses `readTargetOptions` to get the build properties.

## Expected Behavior

Either `context.workspace` should come with implicit target or `readTargetOptions` should invoke the `registerProjectTargets` logic.

## Steps to Reproduce

Create an workspace addon which uses `registerProjectTargets` to define a build and serve for react app:

```typescript
import { TargetConfiguration } from '@nrwl/devkit';

export const projectFilePatterns = ['project.json', 'my-other-project-file'];

export function registerProjectTargets(
  projectFilePath: string
): Record<string, TargetConfiguration> {
  const project = {
    name: 'react-app',
    root:  'apps/react-app',
    root:  'apps/react-app',
  };

  return {
      serve: {
        executor: '@nrwl/web:dev-server',
        options: {
          buildTarget: `${project.name}:build`,
          hmr: true,
          proxyConfig: `${project.root}/proxy.conf.json`,
        },
      },
      build: {
        executor: '@nrwl/web:webpack',
        outputs: ['{options.outputPath}'],
        options: {
          outputPath: `dist/${project.name}`,
          index: `${project.root}/public/index.html`,
          baseHref: '',
          main: `${project.root}/src/index.tsx`,
          polyfills: `${project.root}/src/polyfills.ts`,
          tsConfig: `${project.root}/tsconfig.app.json`,
          assets: [
            {
              input: `./${project.root}/public`,
              glob: '**/*.!(html)',
              output: '.',
            },
          ],
          styles: [],
          scripts: [],
          webpackConfig: 'webpack.config.app.js',
        },
      },
  };
}
```

Then run: `nx run react-app:serve`

### Failure Logs

```
Cannot read property 'executor' of undefined
TypeError: Cannot read property 'executor' of undefined
    at readTargetOptions (workspace\packages\devkit\src\executors\read-target-options.ts:19:58)
    at getBuildOptions (workspace\packages\web\src\executors\dev-server\dev-server.impl.ts:129:41)
    at devServerExecutor_1 (workspace\packages\web\src\executors\dev-server\dev-server.impl.ts:49:5)
    at devServerExecutor_1.next (<anonymous>)
    at resume (workspace\node_modules\@nrwl\web\node_modules\tslib\tslib.js:225:48)
    at workspace\node_modules\@nrwl\web\node_modules\tslib\tslib.js:224:125    at new Promise (<anonymous>)
    at Object.i.<computed> [as next] (workspace\node_modules\@nrwl\web\node_modules\tslib\tslib.js:224:67)
    at \packages\nx\src\command-line\run.ts:87:25
    at Generator.next (<anonymous>)

 —————————————————————————————————————————————————————————————————————————————————————————————

 >  NX   Running target "react-app:serve" failed

   Failed tasks:

   - react-app:serve:development

   Hint: run the command with --verbose for more details.

error Command failed with exit code 1.
```

### Environment

nx@14.3.5

**Comment by JulioC** [comment]
If this is an expected behavior of `readTargetOptions` and `context.workspace`, maybe it should be addressed in the `web` package

**Comment by JulioC** [comment]
I believe this request may have high impact to be fixed, but I'd like to ask for a workaround...

The original request was made to allow the [webpack dev-server executor](https://github.com/nrwl/nx/blob/f49769a34a7f47d252648132793cddd2612262ee/packages/webpack/src/executors/dev-server/dev-server.impl.ts#L111) to be run without defining a `build` target explicitly in the project.json.

Would it be feasible to change the dev-server executor to be runnable without the build target?

**Comment by JulioC** [comment]
@AgentEnder I've went ahead and implemented the change in `readTargetOptions` as proposed in my original message. I believe this is the right approach, as it's aligned with the way you initially implemented the project inference in #8210.

I'm using this change in my projects, and it appears to be working as expected.

**Comment by AgentEnder** [comment]
The suggested fix is valid, but I think there's a better fix. I'll leave a review.
